lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -772,6 +772,65 @@ def accessible_objects(user_or_attribute, operation, options = {})
       end
     end
 
+    # Version of accessible_objects which only returns objects that are
+    # ancestors of a specified root object or the object itself
+    def accessible_ancestor_objects(user_or_attribute, operation, root_object, options = {})
+      # Drop down to the root's stored policy element if not already there
+      root_object_pe = root_object.respond_to?(:stored_pe) ? root_object.stored_pe : root_object
+
+      # The final set of accessible objects must be contained in the following set
+      ancestor_objects = root_object_pe.ancestors(type: class_for_type('object')) + [root_object_pe]
+
+      # Short-circuit and return all ancestors (minus prohibitions), if authorized on the root node
+      if is_privilege?(user_or_attribute, operation, root_object_pe)
+        if options[:ignore_prohibitions] || !(prohibition = prohibition_for(operation))
+          if inclusion = options[:includes]
+            return ancestor_objects.select { |obj| obj.send(options[:key].to_sym).include?(inclusion) }
+          else
+            return ancestor_objects
+          end
+        else
+          ancestor_objects_minus_prohibitions = ancestor_objects - accessible_ancestor_objects(user_or_attribute, prohibition_for(operation), root_object_pe)
```

Comment:
I'm wondering if it's worth it (or works at all) to have an optional ancestor_objects parameter that can be passed to prevent the recursion here from hitting the db again in line#782.